src/Nethermind/Nethermind.Network.Test/MessageDictionaryTests.cs

```diff
@@ -0,0 +1,106 @@
+using System;
+using System.Collections.Generic;
+using FluentAssertions;
+using Nethermind.Core;
+using Nethermind.Core.Test.Builders;
+using Nethermind.Network.P2P;
+using Nethermind.Network.P2P.Subprotocols;
+using Nethermind.Network.P2P.Subprotocols.Eth.V66.Messages;
+using NUnit.Framework;
+using GetBlockHeadersMessage = Nethermind.Network.P2P.Subprotocols.Eth.V62.Messages.GetBlockHeadersMessage;
+
+namespace Nethermind.Network.Test;
+
+public class MessageDictionaryTests
+{
+    private readonly List<Eth66Message<GetBlockHeadersMessage>> _recordedRequests = new();
+    private MessageDictionary<Eth66Message<GetBlockHeadersMessage>, GetBlockHeadersMessage, BlockHeader[]>
+        _testMessageDictionary;
+
+    [SetUp]
+    public void Setup()
+    {
+        _recordedRequests.Clear();
+        _testMessageDictionary = new((message) => _recordedRequests.Add(message));
+    }
+
+    [Test]
+    public void Test_SendAndReceive()
+    {
+        Request<Eth66Message<GetBlockHeadersMessage>, BlockHeader[]> request = CreateRequest(111);
+
+        BlockHeader[] response = new[] { Build.A.BlockHeader.TestObject };
+
+        _testMessageDictionary.Send(request);
+
+        _recordedRequests.Count.Should().Be(1);
+        _recordedRequests[^1].Should().BeSameAs(request.Message);
+        request.CompletionSource.Task.IsCompleted.Should().BeFalse();
+
+        _testMessageDictionary.Handle(111, response, 100);
+        request.CompletionSource.Task.IsCompleted.Should().BeTrue();
+        request.CompletionSource.Task.Result.Should().BeSameAs(response);
+    }
+
+    [Test]
+    public void Test_SendAndReceive_withDifferentRequestId()
+    {
+        Request<Eth66Message<GetBlockHeadersMessage>, BlockHeader[]> request = CreateRequest(111);
+
+        BlockHeader[] response = new[] { Build.A.BlockHeader.TestObject };
+
+        _testMessageDictionary.Send(request);
+
+        _recordedRequests.Count.Should().Be(1);
+        _recordedRequests[^1].Should().BeSameAs(request.Message);
+        request.CompletionSource.Task.IsCompleted.Should().BeFalse();
+
+        _testMessageDictionary.Invoking((dictionary) => dictionary.Handle(112, response, 100))
+            .Should()
+            .Throw<SubprotocolException>();
+    }
+
+    [Test]
+    public void Test_SendAndReceive_outOfOrder()
```

Comment:
At this point we are able to send messages and handle receiving responses out of order - it's awesome! But I think we can go further and try to drop `await` - more context below, in comment to `Eth66ProtocolHandler`